clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package etcdinit
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/clustermesh-apiserver/etcdinit/defaults"
+	einit "github.com/cilium/cilium/pkg/etcd/init"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/spf13/cobra"
+	client "go.etcd.io/etcd/client/v3"
+)
+
+var (
+	log                     = logging.DefaultLogger.WithField(logfields.LogSubsys, "etcdinit")
+	etcdBinaryLocation      = "/usr/bin/etcd"
+	etcdDataDir             = ""
+	etcdClusterName         = ""
+	etcdInitialClusterToken = ""
+	debug                   = false
+)
+
+func NewCmd() *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "etcdinit",
+		Short: "Initialise an etcd data directory for use by the etcd sidecar of clustermesh-apiserver",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return InitEtcdLocal()
+		},
+	}
+	rootCmd.Flags().StringVar(&etcdDataDir, "data-dir", defaults.EtcdDataDir, "etcd data directory")
+	rootCmd.Flags().StringVar(&etcdClusterName, "cluster-name", defaults.EtcdClusterName, "etcd cluster name")
```

Comment:
I'd avoid mixing the Cilium cluster name and etcd cluster name concepts, as it could lead to confusion. 

Please also notice that allowing the etcd cluster name to be changed makes it no longer consistent with that specified in the main etcd container. My assumption is that it doesn't really matter and this parameter should be taken into account only at creation time (needs to be confirmed though), but it could make sense to just hard-code it for the moment to the same value, and then re-evaluate later. The same applies for the initial-cluster-token below.

Reply:
Cluster name is currently passed in to the `clustermesh-apiserver` at runtime via an argument too, as well as being needed for the actual runtime etcd image. So I don't think that if you changed it in Helm it would conflict as it would change all of them? This is the current behaviour.

As for the initial token, my understanding is that this is only really for etcd's clustering. So the only time it would matter is if the end-user has other etcd clusters in the same cluster (or visible in the same subnet as the pods) — to prevent our etcd accidentally getting joined to an outside cluster. I don't think there's an equivalent anywhere in the code because the clients don't use that value. Here we pass it in from the Helm chart and make sure it's the same.

Reply:
> Cluster name is currently passed in to the clustermesh-apiserver at runtime via an argument too, as well as being needed for the actual runtime etcd image. So I don't think that if you changed it in Helm it would conflict as it would change all of them? This is the current behaviour.

Cluster name and etcd cluster name are two separate concepts. The former is the one used by Cilium and configured through the helm chart. In this context, it is used in the init container to form the admin username, and then by the main clustermesh-apiserver containers. The etcd cluster name instead used to be hardcoded to `clustermesh-apiserver` both in the init container and in the main etcd container. 

With this PR, instead, it is set to the same value of the cluster name also in the init container. The long term goal here is to make the etcd cluster name non-deterministic, so that we can distinguish client side when we connected to a different instance of the clustermesh-apiserver. But, for the moment, it could just continue to be hardcoded to `clustermesh-apiserver`, for symmetry with the main etcd container.

> As for the initial token, my understanding is that this is only really for etcd's clustering. So the only time it would matter is if the end-user has other etcd clusters in the same cluster (or visible in the same subnet as the pods) — to prevent our etcd accidentally getting joined to an outside cluster. I don't think there's an equivalent anywhere in the code because the clients don't use that value. Here we pass it in from the Helm chart and make sure it's the same.

Yep, I agree.

Reply:
Ah I see. Now I understand. I could add two flags to the command — one for `--etcd-cluster-name` and one for `--cilium-cluster-name`?

Reply:
As you prefer, hardcoding the etcd one should be fine as well (and the same for the initial token). I'd only suggest calling the cilium flag `--cluster-name` for consistency with all the other usages.

Reply:
Done.